utils/csv2yaml.py

```diff
@@ -0,0 +1,176 @@
+                else:
+                    raise Exception("No Permissible Value")
+            elif SUPPORTED_TYPES[dtp.lower()] != "string":
+                class_dict["attributes"][variable_names[ii]]["range"] = SUPPORTED_TYPES[dtp.lower()]
+            # checking for aliases
+            aliases_l = []
+            if df_row["Proposed BICAN Field"][ii].lower() != variable_names[ii].lower():
+                aliases_l.append(df_row["Proposed BICAN Field"][ii].strip())
+            if df_row["Aliases"][ii]:
+                for el in df_row["Aliases"][ii].split(","):
+                    if el.strip().lower() not in [variable_names[ii].strip().lower(), df_row["Proposed BICAN Field"][ii].strip().lower()]:
+                        aliases_l.append(el.strip())
+            if aliases_l:
+                class_dict["attributes"][variable_names[ii]]["aliases"] = aliases_l
+            # checking for required
+            if not df_row["Nullable"][ii]:
+                class_dict["attributes"][variable_names[ii]]["required"] = True
+
+    # adding slots_usage if they are any predicate from prov
+    if df_row["Predicate"] and not pd.isna(df_row["Class Name"]):
+        class_dict["slot_usage"] = {df_row["Predicate"]:  {"range": df_row["Subject"]}}
+
+    return class_dict
+
+
+def _create_subsets(csv_dirname):
+    """Create a dictionary with subsets"""
+    df_sub = pd.read_csv(csv_dirname / "subsets.csv")
+    # changing float.nan to None
+    df_sub = df_sub.applymap(lambda x: None if pd.isna(x) else x)
+
+    subsets_dict = {}
+    for row in df_sub.iterrows():
+        subsets_dict[row[1]["Name"]] = {}
+        if row[1]["Description"]:
+            subsets_dict[row[1]["Name"]]["description"] = row[1]["Description"]
+
+    return subsets_dict
+
+
+def _create_valuesets(csv_dirname):
+    """Create a dictionary with valuesets"""
+    df_valsets = pd.read_csv(csv_dirname / "valuesets.csv")
+    # changing float.nan to None
+    df_valsets = df_valsets.applymap(lambda x: None if pd.isna(x) else x)
+    dfg_valsets = df_valsets.groupby("ValueSet Name").agg(lambda x: list(x))
+    dfg_valsets['Enum Name'] = dfg_valsets.index
+
+    valsets_dict = {}
+    for row in dfg_valsets.iterrows():
+        enum_nm = row[1]["Enum Name"].title().replace(" ", "")
+        valsets_dict[enum_nm] = {"permissible_values": {}}
+        for ii, val in enumerate(row[1]["Value"]):
+            if row[1]["Value Description"][ii]:
+                valsets_dict[enum_nm]["permissible_values"][val] = {"description": row[1]["Value Description"][ii]}
+            else:
+                valsets_dict[enum_nm]["permissible_values"][val] = None
+    return valsets_dict
+
+
+def csv2yaml(csv_dirname, yaml_filename, templates_dirname=None):
+    """ Convert a csv file to a linkml schema"""
+    csv_dirname = Path(csv_dirname)
+    if templates_dirname:
+        templates_dirname = Path(templates_dirname)
+    for tab_name in ["slots", "classes", "subsets", "relations"]:
```

Comment:
this seems to assume a structure of the CSV dir, instead of the structure in gdoc. is there a separate process that converts gdoc to csv?

Reply:
yes, I do. Adding an automatic reading will be next step, will do it now